packages-exp/auth-exp/src/core/auth/auth_impl.test.ts

```diff
@@ -117,6 +118,102 @@ describe('AuthImpl', () => {
       );
     });
   });
+
+  describe('change listeners', () => {
+    // // Helpers to convert auth state change results to promise
+    // function onAuthStateChange(callback: NextFn<User|null>)
+
+    it('immediately calls authStateChange if initialization finished', done => {
+      const user = testUser('uid');
+      auth.currentUser = user;
+      auth._isInitialized = true;
+      auth.onAuthStateChanged(user => {
+        expect(user).to.eq(user);
+        done();
+      });
+    });
+
+    it('immediately calls idTokenChange if initialization finished', done => {
+      const user = testUser('uid');
+      auth.currentUser = user;
+      auth._isInitialized = true;
+      auth.onIdTokenChange(user => {
+        expect(user).to.eq(user);
+        done();
+      });
+    });
+
+    it('immediate callback is done async', () => {
+      auth._isInitialized = true;
+      let callbackCalled = false;
+      auth.onIdTokenChange(() => {
+        callbackCalled = true;
+      });
+
+      expect(callbackCalled).to.be.false;
+    });
+
+    describe('user logs in/out, tokens refresh', () => {
+      let user: User;
+      let callback: sinon.SinonSpy;
+
+      beforeEach(() => {
+        user = testUser('uid');
+        callback = sinon.spy();
+      });
+
+      it('onAuthStateChange triggers on log in', async () => {
+        await auth.updateCurrentUser(null);
+        auth.onAuthStateChanged(callback);
+        await auth.updateCurrentUser(user);
+        expect(callback).to.have.been.calledTwice;
+      });
+
+      it('onAuthStateChange triggers on log out', async () => {
+        await auth.updateCurrentUser(user);
+        auth.onAuthStateChanged(callback);
+        await auth.updateCurrentUser(null);
+        expect(callback).to.have.been.calledTwice;
+      });
+
+      it('onIdTokenChange triggers on log in', async () => {
+        await auth.updateCurrentUser(null);
+        auth.onIdTokenChange(callback);
+        await auth.updateCurrentUser(user);
+        expect(callback).to.have.been.calledTwice;
+      });
+
+      it('onIdTokenChange triggers on log out', async () => {
+        await auth.updateCurrentUser(user);
+        auth.onIdTokenChange(callback);
+        await auth.updateCurrentUser(null);
+        expect(callback).to.have.been.calledTwice;
```

Comment:
can you move the first callback into setup so these are either have.been.calledOnce or not.have.been.called?

Reply:
Done. Yeah the problem is that the the callback is always immediately triggered. I've just switched to resetting the spy